Python/Module5_OddsAndEnds/Writing_Good_Code.ipynb

```diff
@@ -0,0 +1,1028 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "In the case that you have [defined your own type (a.k.a class)](http://www.pythonlikeyoumeanit.com/module_4.html), you can simply provide the resulting class object in an annotation. This type-hints that a variable is expected to be an *instance* of that type. If, instead, you have the rare case of hinting that the class object itself (not an instance of the class) is expected, you can use `typing.Type` to indicate this.\n",
+    "\n",
+    "Let's write a silly example to demonstrate these points. We will define our own `Dog` class, and will write a function that expects the `Dog` class-object itself (type-hinted as `Type[Dog]`). The function will create several *instances* of the `Dog` class (type-hinted simply as `Dog`), and will return them in a list."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "```python\n",
+    "# type-hinting a custom class\n",
+    "from typing import List, Type\n",
+    "class Dog:\n",
+    "    def __init__(self, name):\n",
+    "        self.name = name\n",
+    "\n",
+    "# cls is expected to be the class-object, `Dog`, itself\n",
+    "# This function returns a list of instances of the `Dog` type \n",
+    "def list_famous_dogs(cls: Type[Dog]) -> List[Dog]:\n",
+    "    return [cls(name) for name in [\"Lassie\", \"Shadow\", \"Air Bud\"]]\n",
+    "```\n",
+    "\n",
+    "You can also create a [type alias](https://docs.python.org/3/library/typing.html#type-aliases), which you can define in your code and use in your annotations. For example, if you will regularly be passing around a tuple containing five `Dog`-instances, you can define an alias to help make your type-hints more succinct:\n",
+    "\n",
+    "```python\n",
+    "# creating an alias for a type\n",
+    "Pack = Tuple[Dog, Dog, Dog, Dog, Dog]\n",
+    "\n",
+    "def find_alpha(dogs: Pack) -> Dog:\n",
+    "    ...\n",
+    "```\n",
+    "\n",
+    "Object types defined in 3rd-party libraries, e.g. NumPy's [ndarray](https://www.pythonlikeyoumeanit.com/Module3_IntroducingNumpy/IntroducingTheNDarray.html) behave no differently from our own custom custom classes; simply provide the [class object](https://www.pythonlikeyoumeanit.com/Module4_OOP/ClassDefinition.html#Defining-a-New-Class-of-Object) in the type-hint annotation:\n",
+    "\n",
+    "```python\n",
+    "# type-hinting an nd-array from numpy\n",
+    "import numpy as np\n",
+    "\n",
+    "def custom_dot_product(x: np.ndarray, y: np.ndarray) -> float:\n",
+    "    return float(np.sum(x * y))\n",
+    "```\n",
+    "\n",
+    "\n",
+    "Eventually popular 3rd party libraries like NumPy will contribute their own typing modules so that you can provide higher-fidelity hints that indicate things like data-type and array-shape. NumPy developers [are currently working on this](https://github.com/numpy/numpy/issues/7370)."
```

Comment:
Eventually**,** popular 3rd...

Otherwise you're talking about those third-party libraries that are eventually popular